src/relay/pass/partial_eval.cc

```diff
@@ -232,12 +233,145 @@ struct SFuncNode : StaticNode {
+   *
+   * Thus, if the generated is smaller then the argument of Meet,
+   * and the generated is not smaller then (*this),
+   * progress should be false.
+   */
+  virtual std::tuple<Fuel, bool> Meet(const Fuel& f) const {
+    bool progress = false;
+    auto ret = Meet(f, &progress);
+    return std::make_tuple(ret, progress);
+  }
+  /*! \brief return the new Fuel, and write true only iff progress is made. */
+  virtual Fuel Meet(const Fuel& f, bool* progress) const {
+    CHECK(progress);
+    auto ret = Meet(f);
+    *progress |= std::get<1>(ret);
+    return std::get<0>(ret);
+  }
+  static constexpr const char* _type_key = "relay.Fuel";
+  TVM_DECLARE_BASE_NODE_INFO(FuelNode, RelayNode);
+};
+
+const FuelNode* Fuel::operator->() const {
+  return static_cast<const FuelNode*>(node_.get());
+}
+
+Fuel MkFSeq(const std::vector<Fuel>& fuels);
+struct FSeqNode : FuelNode {
+  std::vector<Fuel> fuels;
+  Fuel Meet(const Fuel& f, bool* progress) const final {
+    auto x = f.as<FSeqNode>();
+    CHECK(x);
+    CHECK_EQ(fuels.size(), x->fuels.size());
+    std::vector<Fuel> new_fuels;
+    for (size_t i = 0; i < fuels.size(); ++i) {
+      new_fuels.push_back(fuels[i]->Meet(x->fuels[i], progress));
+    }
+    return MkFSeq(new_fuels);
+  }
+  explicit FSeqNode(const std::vector<Fuel>& fuels) : fuels(fuels) { }
+  static constexpr const char* _type_key = "relay.FSeq";
+  TVM_DECLARE_NODE_TYPE_INFO(FSeqNode, FuelNode);
+};
+
+RELAY_DEFINE_NODE_REF(FSeq, FSeqNode, Fuel);
+
+Fuel MkFSeq(const std::vector<Fuel>& fuels) {
+  return Fuel(make_node<FSeqNode>(fuels));
+}
+
+Fuel MkFTime(Time time);
+struct FTimeNode : FuelNode {
+  Time time;
+  std::tuple<Fuel, bool> Meet(const Fuel& f) const final {
+    auto x = f.as<FTimeNode>();
+    CHECK(x);
+    Time new_time = std::min(time, x->time);
+    return std::make_tuple(MkFTime(new_time), new_time < time);
+  }
+  explicit FTimeNode(Time time) : time(time) { }
+  static constexpr const char* _type_key = "relay.FTime";
+  TVM_DECLARE_NODE_TYPE_INFO(FTimeNode, FuelNode);
+};
+
+RELAY_DEFINE_NODE_REF(FTime, FTimeNode, Fuel);
+
+Fuel MkFTime(Time time) {
+  return Fuel(make_node<FTimeNode>(time));
+}
+
+Fuel MkFTValue(size_t tvalue);
+/*! \brief If the pstatic is hold a positive integer scalar, that number, else 0. */
+struct FTValueNode : FuelNode {
+  size_t tvalue;
+  std::tuple<Fuel, bool> Meet(const Fuel& f) const final {
+    auto x = f.as<FTValueNode>();
+    CHECK(x);
+    size_t new_tvalue = std::min(tvalue, x->tvalue);
+    return std::make_tuple(MkFTValue(new_tvalue), new_tvalue < tvalue);
+  }
+  explicit FTValueNode(size_t tvalue) : tvalue(tvalue) { }
+  static constexpr const char* _type_key = "relay.FTValue";
+  TVM_DECLARE_NODE_TYPE_INFO(FTValueNode, FuelNode);
+};
+
+RELAY_DEFINE_NODE_REF(FTValue, FTValueNode, Fuel);
+
+Fuel MkFTValue(size_t tvalue) {
+  return Fuel(make_node<FTValueNode>(tvalue));
+}
+
+/*! \brief Initially every element has Fuel of FBottom. It is the smallest element.
```

Comment:
I don't understand the code or lattices very well yet, but if every element starts at `FBottom`, and you only do meets, then wouldn't it remain `FBottom` forever, since the meet of a minimal element and any other element is always the minimal element?

If that's the case, then maybe this should be `FTop`?

Reply:
Good catch! It is indeed FTop. It was initally 'finitely ascending' but that sound strange, so I flip the lattice to match that.